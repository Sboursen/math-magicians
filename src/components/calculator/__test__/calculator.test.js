import React from 'react';
import {
  fireEvent,
  render,
  screen,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import Calculator from '../calculator';

describe('Calculator', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('contains exactly 19 buttons', () => {
    render(<Calculator />);

    const buttons = screen.getAllByRole('button');

    expect(buttons.length).toBe(19);
  });

  it('returns the correct result after the 1 button is clicked', () => {
    render(<Calculator />);
    const resultElement = screen.getByTestId(/result element/);
    const OneButton = screen.getByText(/1/);

    fireEvent.click(OneButton);

    expect(resultElement).toHaveTextContent(1);
  });

  it('returns the correct result after a {2, +, 2, =} operation', () => {
    render(<Calculator />);
    const resultElement = screen.getByTestId(/result element/);
    const twoButton = screen.getByText(/2/);
    const plusButton = screen.getByText(/^[+]$/);
    const equalButton = screen.getByText(/=/);

    fireEvent.click(twoButton);
    fireEvent.click(plusButton);
    fireEvent.click(twoButton);
    fireEvent.click(equalButton);

    expect(resultElement).toHaveTextContent(4);
  });

  it('returns an error rafter a {3, รท, 0, =} operation', () => {
    render(<Calculator />);
    const resultElement = screen.getByTestId(/result element/);
    const threeButton = screen.getByText(/3/);
    const zeroButton = screen.getByText(/0/);
    const dividedByButton = screen.getByText(/^[รท]$/);
    const equalButton = screen.getByText(/=/);

    fireEvent.click(threeButton);
    fireEvent.click(dividedByButton);
    fireEvent.click(zeroButton);
    fireEvent.click(equalButton);

    expect(resultElement).toHaveTextContent(
      "Can't divide by 0.",
    );
  });

  it('returns an error rafter a {3, %, 0, =} operation', () => {
    render(<Calculator />);
    const resultElement = screen.getByTestId(/result element/);
    const threeButton = screen.getByText(/3/);
    const zeroButton = screen.getByText(/0/);
    const modeButton = screen.getByText(/^[%]$/);
    const equalButton = screen.getByText(/=/);

    fireEvent.click(threeButton);
    fireEvent.click(modeButton);
    fireEvent.click(zeroButton);
    fireEvent.click(equalButton);

    expect(resultElement).toHaveTextContent(
      "Can't find modulo as can't divide by 0.",
    );
  });
});
