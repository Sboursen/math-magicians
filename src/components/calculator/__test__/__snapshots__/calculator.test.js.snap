// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator renders correctly 1`] = `
<div
  className="calculator-wrap"
>
  <div
    data-testid="result element"
    id="res"
  />
  <button
    className="main "
    onClick={[Function]}
    type="button"
  >
    AC
  </button>
  <button
    className="main "
    onClick={[Function]}
    type="button"
  >
    +/-
  </button>
  <button
    className="main "
    onClick={[Function]}
    type="button"
  >
    %
  </button>
  <button
    className="aside "
    onClick={[Function]}
    type="button"
  >
    รท
  </button>
  <button
    className="main "
    onClick={[Function]}
    type="button"
  >
    7
  </button>
  <button
    className="main "
    onClick={[Function]}
    type="button"
  >
    8
  </button>
  <button
    className="main "
    onClick={[Function]}
    type="button"
  >
    9
  </button>
  <button
    className="aside "
    onClick={[Function]}
    type="button"
  >
    x
  </button>
  <button
    className="main "
    onClick={[Function]}
    type="button"
  >
    4
  </button>
  <button
    className="main "
    onClick={[Function]}
    type="button"
  >
    5
  </button>
  <button
    className="main "
    onClick={[Function]}
    type="button"
  >
    6
  </button>
  <button
    className="aside "
    onClick={[Function]}
    type="button"
  >
    -
  </button>
  <button
    className="main "
    onClick={[Function]}
    type="button"
  >
    1
  </button>
  <button
    className="main "
    onClick={[Function]}
    type="button"
  >
    2
  </button>
  <button
    className="main "
    onClick={[Function]}
    type="button"
  >
    3
  </button>
  <button
    className="aside "
    onClick={[Function]}
    type="button"
  >
    +
  </button>
  <button
    className="main wide-button"
    onClick={[Function]}
    type="button"
  >
    0
  </button>
  <button
    className="main "
    onClick={[Function]}
    type="button"
  >
    .
  </button>
  <button
    className="aside "
    onClick={[Function]}
    type="button"
  >
    =
  </button>
</div>
`;
